Allow setting alternator_enforce_authorization to "warn"

nyh opened the pull request:
An Alternator user was recently "bit" when switching `alternator_enforce_authorization` from "false" to "true": ְְְAfter the configuration change, all application requests suddenly failed because unbeknownst to the user, their application used incorrect secret keys. 

This series introduces a solution for users who want to **safely** switch `alternator_enforce_authorization`  from "false" to "true": Before switching from "false" to "true", the user can temporarily switch to "warn". In "warn" mode, authentication and authorization errors are counted in metrics (`scylla_alternator_authentication_failures` and `scylla_alternator_authorization_failures`) but the user's application continues to work. The user can then fix any errors in their application's setup, and only do the switch to "true" when the metrics show there are no more errors.

The first patch is the feature itself (the "warn" mode and the metrics), the second patch is a test for it, and the third patch is documentation recommending how to use the "warn" mode and the associated metrics to safely switch from "false" to "true".

Fixes #25308

This is a feature that users need, so it should probably be backported to live branches.

nyh commented on alternator/server.cc:
```diff
@@ -342,29 +359,37 @@ future<std::string> server::verify_signature(const request& req, const chunked_c
-    });
+        throw api_error::unrecognized_client("The security token included in the request is invalid.");
+    }
+    co_return user;
```
Yes, but before coroutinization the code was super-ugly and almost impossible to write correctly. I can probably split this into a separate patch that co-routinizes before changing anything, just to make it easier to review, if you're finding it hard to review like this (it's only a small patch of code anyway, so I thought it would be ok even without a separate patch).

amnonh replied:
I suggest adding a log line with more specific details if they are available. I believe/hope that in the end we'll end up with logs instead of metrics in this case.

amnonh replied:
Getting log lines to the monitoring is my current windmill-tilting. I'm optimistic.

Logs can serve just as a good data source. They carry specific information with them that metrics do not. You can create graphs, alerts, and even metrics if needed. 
I think our logs are in a sub-optimal state because we don't use them well enough.

nyh replied:
> I suggest adding a log line with more specific details if they are available. I believe/hope that in the end we'll end up with logs instead of metrics in this case.

I agree that having log messages that tell you exactly which client had which kind of error can be useful in case not all your clients are misconfigured. We just need to be careful not to log too much sensitive authentication information, as this in itself will become a security problem. For example, obviously we can't log the secret key used in the failed authentication. So if the user's problem is sending a wrong secret key, we still can't show the user in the log which incorrect key was received. We can, however, say the client's IP address and perhaps the username attempted to authenticate.

nyh replied:
We should avoid having different discussions on the same review comment, it makes it difficult to understand what review comments were addressed and which were not :-(

Anyway, I'm working now to add the log messages in addition to the metrics:
1. Since the log messages are intentional and important - to be documented and expected by users to work - I'll need to also write a test for them, not just the metrics.
2. I think the log messages should appear **only** when alternator_enforce_authorization is set to "warn" - when it's set to "true", clients will get errors directly and don't need anything in the log. Please tell me if you don't agree with me on this.
3. For the same reason the log messages will have "WARN" level - when the user sets alternator_enforce_authorization to "warn", he really wants to see those messages and won't want them to be hidden as some sort of debug-level messages.
4. The log message should contain a bit of information to help identify the problematic client - probably its IP address and username (which I think is benign?), but obviously not the password.

nyh replied:
I'm resolving this discussion. It's about two things - logging, which I added, and coroutinization, which I cancelled in the new version.

ScyllaPiotr replied:
> Another reviewer also asked to have on the dashboard a warning if this metric is ever more than zero

Before scanning the changes, would this warning cleverly indicate that there were # failed authorizations etc. over last 1h, 24h or 7d, or would it dumbly sum all failed logins there happened since a node or cluster started?

ScyllaPiotr replied:
> 2\. I think the log messages should appear **only** when alternator_enforce_authorization is set to "warn" - when it's set to "true", clients will get errors directly and don't need anything in the log. Please tell me if you don't agree with me on this.

I propose to have the log messages present also when `alternator_enforce_authorization` is set to `true`. It would be unexpected for the users to see the log messages disappear with `warn`->`true`. Also, it is beneficial to see these messages in the logs security-wise - a DB admin would like to know if there were some foreign actors trying to authorize to Alternator, as it would be indicative to misconfiguration of the security policy of the network.

amnonh commented on alternator/stats.cc:
```diff
@@ -155,6 +155,15 @@ static void register_metrics_with_optional_table(seastar::metrics::metric_groups
+                seastar::metrics::description("total number of authentication failures"), labels),
+            seastar::metrics::make_counter("authorization_failures", stats.authorization_failures,
+                seastar::metrics::description("total number of authorization failures"), labels),
+        });
```
I would add it in an existing ```if``` next to op_latency_summary is a good candidate.

I think we can report it once per node, no need to do it per shard.
And there's no need to report it if there were never any problems so

so add ```.aggregate(seastar::metrics::shard_label).set_skip_when_empty()``` to the end of ```seastar::metrics::make_counter(....)``

You can use the labels parameter like you did so it would look like the rest of the code.

It will be nice to add a few words about that failure. it's obvious to someone that now it it means, but it's not for everyone else

nyh replied:
If can move it there - I looked for a "good" place and didn't find one.
Why do we stick aggregate(shard_label) on **every** metric, not just big histograms or multi-labeled counters (e.g., a separate counter for each of 20 different types of operation)? Is this the new normal? And same for set_skip_when_empty() - I remember @avikivity said he found skipped metrics confusing - so is it worth doing this for every small counter we add?

What do you mean "add a few words about that failure"? Do you mean add more context in the cover letter about the problem the failure that the user had? I thought I wrote it fairly clearly (in the cover letter). Or do you mean the metrics' description string is missing something? Or what?

amnonh replied:
typically, we don't aggregate over specific metrics, but in this case, I don't see the point in reporting this metric per shard. I'll not insist if you think it's useful.

When we report metrics per table, we do aggregate over shards because the cardinality is too high.
Using skip_when_empty and aggregated histograms helped reduce the number of reported metrics by an order of magnitude.
The skip when empty doesn't only mean the value is zero, it means it was never used, it's just lazy evaluation, not different than how all alternator reporting works.
If everyone were as thoughtful as you when you waited to register the metrics to the time they are needed. We would have been in a much better place.

Add a few words for the description. If someone looks at a metric and the only thing they see is "total number of authorization failures," I'm not sure how easy it will be to understand.

One last thing, please add ```_total``` as a suffix for the metrics names. It complies with Prometheus naming convention, and I'm trying to get new names up to code.

nyh replied:
> I would add it in an existing `if` next to op_latency_summary is a good candidate.

I don't understand what you mean :-( Maybe you meant the if() around CreateTable/DeleteTable, but there the whole idea was to have it next to the other operations - here we're adding completely different metrics so I thought it should be in a different place in the code.

> I think we can report it once per node, no need to do it per shard. And there's no need to report it if there were never any problems so
> 
> so add `.aggregate(seastar::metrics::shard_label).set_skip_when_empty()` to the end of `seastar::metrics::make_counter(....)`

Done. Please review the next version I pushed if I did it correctly (it passes my test, so the metric works at the node level, but maybe something's off or inefficient).

> 
> You can use the labels parameter like you did, so it looks like the rest of the code.

I have absolutely no idea what the "labels" thing does :-( Should I keep it? 

> 
> It will be nice to add a few words about that failure. It's obvious to someone who knows what it means, but it's not for everyone else

amnonh reviewed (changes requested):
In general, it looks good. I've mentioned a few minor things

nyh commented:
I think the CI failure is a flaky test that can't properly handle the encryption key host (?) being temporarily down. Opened https://github.com/scylladb/scylladb/issues/25481.

susiesolis commented:
Customer is asking for an ETA on this setting.

mykaul commented:
> Customer is asking for an ETA on this setting.

@susiesolis - customer issues are not discussed here.

nyh commented:
I think I addressed most review comments (tell me if I missed something). The coroutinization of one piece of code is now moved to a separate first patch, and a fifth patch adds log messages (and extensive tests for them) for auth errors in "warn" mode - in addition to the metrics I already had. I'm personally NOT convinced that these log messages are a good idea, but this was asked by reviewers, and keeping the logging in a separate patch will make it easy to rethink later.

avikivity commented:
@yaronkaikov what's needed to get a CI run here?

avikivity commented on alternator/executor.hh:
```diff
@@ -24,6 +24,7 @@
 #include "utils/updateable_value.hh"
 
 #include "tracing/trace_state.hh"
+#include "db/config.hh"
```
Please don't add db/config.hh to headers. It amplifies changes - any component can change db/config.hh, and then alternator has to be rebuilt.

nyh replied:
We're already pretty-much screwed in that respect - even if header files don't include db/config.hh, still ever source file that needs anything from the configuration includes it. A quick grep counted 126 (!) such source files.

The reason why I found myself including it in a **header** file is that I need the type db::tri_mode_restriction in that header file (as a parameter type), and that type is defined in db/config.hh. I can fix this by moving tri_mode_restriction to its own header file, and include it in db/config.hh - and here in alternator/executor.hh. I'll do that.

nyh commented on alternator/server.cc:
```diff
@@ -342,29 +342,20 @@ future<std::string> server::verify_signature(const request& req, const chunked_c
-            throw api_error::invalid_signature(e.what());
-        }
+    key_cache::value_ptr key_ptr = co_await _key_cache.get_ptr(user, cache_getter);
+    std::string signature;
```
Done. The next version I'll push will no longer have the coroutinization.

nyh commented:
Pushed a new version addressing reviewer comments:
1. Don't coroutinize server::verify_signature(). While the coroutinized version was slightly nicer, it incurs a (small) additional overhead in the typical case that the user's information is already cached (or when auth checks are turned off).
2. Make the new metrics disappear when zero, and aggregated per-node instead of per-shard. I'm not convinced this is better, but Amnon asked for this, so I obliged.

nyh commented:
>     * [test_clear_gently_foreign_shared_ptr_const_payload.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/19208/testReport/junit/boost.stall_free_test/cc/debug___test_py___debug___test_clear_gently_foreign_shared_ptr_const_payload_debug_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_clear_gently_foreign_shared_ptr_const_payload.debug.2)

I think this is a flaky test, previously-unknown - it's not caused by or related to this patch. 
Opened an issue https://github.com/scylladb/scylladb/issues/25723.
> 
>     * [test_basic_restore_with_kms_rotate_key](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/19208/testReport/junit/manager_restore_tests/TestRestoreWithEaR/release___dtest_with_tablets___test_basic_restore_with_kms_rotate_key) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_basic_restore_with_kms_rotate_key)
> 
>     * [test_basic_restore_with_kms_rotate_key](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/19208/testReport/junit/manager_restore_tests/TestRestoreWithEaR/release___dtest_with_gossip_topology_changes___test_basic_restore_with_kms_rotate_key) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_basic_restore_with_kms_rotate_key)
> 
>     * [test_basic_restore_with_kms_rotate_key](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/19208/testReport/junit/manager_restore_tests/TestRestoreWithEaR/release___dtest_with_consistent_topology_changes___test_basic_restore_with_kms_rotate_key) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_basic_restore_with_kms_rotate_key)
> 
>     * [test_basic_restore_with_kms_rotate_key](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/19208/testReport/junit/manager_restore_tests/TestRestoreWithEaR/release___dtest_with_tablets___test_basic_restore_with_kms_rotate_key_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_basic_restore_with_kms_rotate_key)
> 
>     * [test_basic_restore_with_kms_rotate_key](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/19208/testReport/junit/manager_restore_tests/TestRestoreWithEaR/release___dtest_with_consistent_topology_changes___test_basic_restore_with_kms_rotate_key_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_basic_restore_with_kms_rotate_key)
> 
>     * [test_basic_restore_with_kms_rotate_key](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/19208/testReport/junit/manager_restore_tests/TestRestoreWithEaR/release___dtest_with_gossip_topology_changes___test_basic_restore_with_kms_rotate_key_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_basic_restore_with_kms_rotate_key)

A flaky test, https://github.com/scylladb/scylladb/issues/25481

> 
> 
> #### Build Details:
> 
>     * Duration: 6 hr 53 min

nyh commented:
Unfortunately my latest changes (to revert from coroutines to continuations, following Avi's review) broke something real in the test. I'll fix and send a new version.

nyh commented:
Indeed, the Scylla manager test  test_basic_restore_with_kms_rotate_key failed with (see https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/19392/testReport/junit/manager_restore_tests/TestRestoreWithEaR/release___dtest_with_gossip_topology_changes___test_basic_restore_with_kms_rotate_key/)
```
Error: create backup target: create cluster session: no RPC addresses found; node info: giving up after 2 attempts: agent [HTTP 500] unexpected error, consult logs: 
node info fetch: agent: fetch Scylla config 1: json: cannot unmarshal string into Go value of type bool; node info: giving up after 2 attempts: agent [HTTP 500] unexpected error, consult logs:
node info fetch: agent: fetch Scylla config 1: json: cannot unmarshal string into Go value of type bool
Trace ID: mwW0N0ggQwy6ycHFM5ITPw (grep in scylla-manager logs)
```

I don't know what to do about this, and may need advice from the Scylla Manager and Scylla Cloud developers how or why they read or write the alternator_enforce_authorization configuration parameter.

mykaul commented:
> Indeed, the Scylla manager test test_basic_restore_with_kms_rotate_key failed with (see https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/19392/testReport/junit/manager_restore_tests/TestRestoreWithEaR/release___dtest_with_gossip_topology_changes___test_basic_restore_with_kms_rotate_key/)
> 
> ```
> Error: create backup target: create cluster session: no RPC addresses found; node info: giving up after 2 attempts: agent [HTTP 500] unexpected error, consult logs: 
> node info fetch: agent: fetch Scylla config 1: json: cannot unmarshal string into Go value of type bool; node info: giving up after 2 attempts: agent [HTTP 500] unexpected error, consult logs:
> node info fetch: agent: fetch Scylla config 1: json: cannot unmarshal string into Go value of type bool
> Trace ID: mwW0N0ggQwy6ycHFM5ITPw (grep in scylla-manager logs)
> ```
> 
> I don't know what to do about this, and may need advice from the Scylla Manager and Scylla Cloud developers how or why they read or write the alternator_enforce_authorization configuration parameter.

@karol-kokoszka ?

amnonh commented on alternator/server.cc:
```diff
@@ -374,6 +378,7 @@ future<std::string> server::verify_signature(const request& req, const chunked_c
         } catch (const api_error& e) {
             _executor._stats.authentication_failures++;
             if (_enforce_authorization.get() == db::tri_mode_restriction_t::mode::WARN) {
```
If I understand correctly, you don't print to the log when there's actually an error.
Don't hold the merge because of it. But if you have a new version, I suggest adding it.

Logs are a good way to show the history of what happened.

nyh replied:
Added this feature (logging even if when an errror would be generated) to the new version. Basically the new version (explained below) allows you to set `alternator_enforce_authorization=true` to generate errors and also set `alternator_warn_authorization=true` to additionally add WARN-level log messages.

nyh commented on alternator/server.cc:
```diff
@@ -342,25 +371,52 @@ future<std::string> server::verify_signature(const request& req, const chunked_c
     };
-    return _key_cache.get_ptr(user, cache_getter).then([this, &req, &content,
+    return _key_cache.get_ptr(user, cache_getter).then_wrapped([this, &req, &content,
                                                     user = std::move(user),
                                                     host = std::move(host),
                                                     datestamp = std::move(datestamp),
                                                     signed_headers_str = std::move(signed_headers_str),
                                                     signed_headers_map = std::move(signed_headers_map),
                                                     region = std::move(region),
                                                     service = std::move(service),
-                                                    user_signature = std::move(user_signature)] (key_cache::value_ptr key_ptr) {
+                                                    user_signature = std::move(user_signature)] (future<key_cache::value_ptr> key_ptr_fut) {
```
Lambdas that are long and complex are the hallmark of continuation-based code :-( In a previous version of this patch, I coroutinized this code but a reviewer asked me to switch back because of performance considerations. So let's not change this right now. Remember that I'm changing pre-existing code.

nyh commented on test/alternator/test_logs.py:
```diff
@@ -0,0 +1,273 @@
+    port2hex = lambda port: f'{int(port):04X}'
+    try:
+        addr1 = ip2hex(ip) + ':' + port2hex(port)
+    except:
```
I know, and I think that for this use - parsing of a single string - what I did is good enough.

nyh commented:
Pushed a new version where the repeated boilerplate code to increment a counter, maybe log and maybe throw an exception was replaced by one function (actually, two - one in server.cc and a slightly different one in executor.cc).
This address two reviewer comments:
1. copilot asked for this repetitive code to be replaced by a function.
2. @ScyllaPiotr asked for it to be even more sophisticated - when warn=true but enforce=false the log message should be slightly different, saying it would have been an error only if enforce will be set to true.

ScyllaPiotr commented on alternator/server.cc:
```diff
@@ -346,26 +382,40 @@ future<std::string> server::verify_signature(const request& req, const chunked_c
+            authentication_error(_executor._stats, _enforce_authorization.get(), _warn_authorization.get(),
+                e, user, req.get_client_address());
+            return std::string();
+        }
```
Ah, with the generic naming (`key_cache::value_ptr`) I associated it with Szymon's request caching, although it would make little sense to expect caching of requests inside `server::verify_signature()`. Thanks for clarifying.

ScyllaPiotr commented on alternator/server.cc:
```diff
@@ -346,26 +382,40 @@ future<std::string> server::verify_signature(const request& req, const chunked_c
             _key_cache.remove(user);
-            throw api_error::unrecognized_client("The security token included in the request is invalid.");
+            authentication_error(_executor._stats, _enforce_authorization.get(), _warn_authorization.get(),
+                api_error::unrecognized_client("wrong signature"),
```
So let's not use "security token", leaving the rest of the message unchanged. The current proposal misses an important word "request".

ScyllaPiotr commented on test/alternator/test_logs.py:
```diff
@@ -0,0 +1,273 @@
+# retries until a given timeout. The timeout may be long, because successful
+# tests will not wait for it. Note, however, that long timeouts will make
+# xfailing tests slow.
+def wait_for_log(logfile, pattern, re_flags=re.MULTILINE, timeout=5):
```
It's questionable if we should prefer coherence over communicability.
To have to look into how the function is called to understand its meaning is the definition of failure of naming.
But that's not very important TBH.

ScyllaPiotr commented on alternator/server.cc:
```diff
@@ -346,26 +382,40 @@ future<std::string> server::verify_signature(const request& req, const chunked_c
         return get_key_from_roles(proxy, as, std::move(username));
     };
-    return _key_cache.get_ptr(user, cache_getter).then([this, &req, &content,
+    return _key_cache.get_ptr(user, cache_getter).then_wrapped([this, &req, &content,
```
Ah, catching exceptions by `then_wrapped()` refers to exceptions from the function earlier in the chain, not from the lambda supplied as a param to `then_wrapped()`. Now I get that.

nyh commented:
> With no merge conflict, there's no real reason to hold this back.

Now there are merge conflicts :-( I'll rebase.

nyh commented:
ping. This is a P1 (!) issue with a pull request which is now over 3 months old, let's not drop the ball.

nyh commented:
> Tried to queue but it conflicts with another alternator PR which is in next. Will need a rebase once promotion finishes.

Thanks. Rebased.

nyh commented:
ping (this is a P1 issue).
